coalib/bearlib/__init__.py

```diff
@@ -82,34 +97,57 @@ def _deprecate_decorator(func):
-                        logged_deprecated_args.add(deprecated_arg)
-                    depr_arg_value = _func.__call__(kwargs[deprecated_arg])
-                    if arg in kwargs and depr_arg_value != kwargs[arg]:
-                        logging.warning(
-                            'The value of `{}` and `{}` are conflicting.'
-                            ' `{}` will be used instead.'.format(
-                                  deprecated_arg, arg, arg))
+            for arg, depr_args_and_modifier in depr_args.items():
+                if isinstance(depr_args_and_modifier, tuple):
+                    if callable(depr_args_and_modifier[-1]):
+                        deprecated_args = depr_args_and_modifier[:-1]
+                        _func = depr_args_and_modifier[-1]
                     else:
-                        kwargs[arg] = depr_arg_value
-                    del kwargs[deprecated_arg]
+                        deprecated_args = depr_args_and_modifier
+                        _func = _do_nothing
+                else:
+                    deprecated_args = (depr_args_and_modifier, )
+                    _func = _do_nothing
+
+                for deprecated_arg in deprecated_args:
+                    if isinstance(deprecated_arg, dict):
+                        modifier = list(deprecated_arg.values())[0]
+                        deprecated_arg = list(deprecated_arg.keys())[0]
+                    else:
+                        modifier = _func
+                    if deprecated_arg in kwargs:
+                        if deprecated_arg not in logged_deprecated_args:
+                            logging.warning(
+                                'The setting `{}` is deprecated. Please use'
+                                ' `{}` instead.'.format(deprecated_arg, arg))
+                            logged_deprecated_args.add(deprecated_arg)
+                        depr_arg_value = modifier.__call__(
+                            kwargs[deprecated_arg])
+                        if arg in kwargs and depr_arg_value != kwargs[arg]:
+                            logging.warning(
+                                'The value of `{}` and `{}` are conflicting.'
+                                ' `{}` will be used instead.'.format(
+                                    deprecated_arg, arg, arg))
+                        else:
+                            kwargs[arg] = depr_arg_value
+                        del kwargs[deprecated_arg]
+
             return func(*args, **kwargs)
 
         new_metadata = FunctionMetadata.from_function(func)
-        for arg, depr_arg_and_modifier in depr_args.items():
-            deprecated_arg = (depr_arg_and_modifier[0]
-                              if isinstance(depr_arg_and_modifier, tuple)
-                              else depr_arg_and_modifier)
-            new_metadata.add_deprecated_param(arg, deprecated_arg)
+        for arg, depr_args_and_modifier in depr_args.items():
```

Comment:
this section of code should be the only code which looks at `depr_args`.
It should save whatever is needed into `__metadata__`, possibly with new fields added.

This chunk of code is run once -- **everything** *possible* should be done here to prepare the decoration, so that each invocation of `wrapping_function` is as fast as possible.

Rename `arg, depr_args_and_modifier` to `arg, depr_value` as its contents & datatype is now not known here